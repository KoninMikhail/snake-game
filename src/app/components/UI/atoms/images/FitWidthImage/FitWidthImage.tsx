import { StyledFitWidthImage } from './style/StyledFitWidthImage';
import useProgressiveImageLoader from '@hooks/useProgressiveImageSource/useProgressiveImageSource';
import { IImage } from '@ui/atoms/images/images.interface';
import { useImagesContext } from '@context/ImagesContext';
import useAppSelector from '@hooks/useAppSelector/useAppSelector';
import { appDeviceInfoSelector } from '@store/slices/app.slice';
import getImageSourceById from '@/helpers/getters/getImageSourceById';
import { DeviceScreenSize } from '@/types/device';

/**
 * @name FitWidthImage
 * @author Mikhail Konin <dev.konin@gmail.com>
 *
 * Use for logo state the page.
 *
 * @param id - image id in context
 * @param alt - logo alt attrib.
 */
export const FitWidthImage = ({ sourceID, alt, ...rest }: IImage) => {
    const context = useImagesContext();
    const deviceInfo = useAppSelector(appDeviceInfoSelector);
    const deviceScreenSize = deviceInfo.screen as DeviceScreenSize;

    const dataForLoading = getImageSourceById(context, sourceID, deviceScreenSize);
    const output = useProgressiveImageLoader(dataForLoading);

    return <StyledFitWidthImage src={output} alt={alt} {...rest} />;
};
