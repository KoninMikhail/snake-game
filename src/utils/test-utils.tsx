import React, {PropsWithChildren} from 'react';
import type {RenderOptions} from '@testing-library/react';
import {render} from '@testing-library/react';
import type {PreloadedState} from '@reduxjs/toolkit';
import {configureStore} from '@reduxjs/toolkit';
import {Provider} from 'react-redux';

import type {AppStore, RootState} from '@store/store';
import appSlice from '@store/slices/app.slice';
import soundSlice from '@store/slices/sound.slice';
import maxScoreSlice from '@store/slices/maxScore.slice';
import gameDataSlice from '@store/slices/game.slice';
import {ImagesData} from '@data/imagesData';
import {mockImagesData, mockSoundData} from '@/setupTests';
import SoundsProvider from '@context/SoundsContext';
import ImagesProvider from '@context/ImagesContext';
import {SoundsData} from '@data/soundsData';
import {DefaultTheme, ThemeProvider} from 'styled-components';
import {theme as appTheme} from '@styles/theme';

// As a basic setup, import your same slice reducers

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: PreloadedState<RootState>;
    store?: AppStore;
    images?: ImagesData;
    sounds?: SoundsData;
    theme?: DefaultTheme;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = configureStore({
            reducer: {
                app: appSlice,
                sounds: soundSlice,
                maxScore: maxScoreSlice,
                game: gameDataSlice,
            },
            preloadedState,
        }),
        images = mockImagesData,
        sounds = mockSoundData,
        theme = appTheme,
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({children}: PropsWithChildren): JSX.Element {
        return (
            <Provider store={store}>
                <ThemeProvider theme={theme}>
                    <SoundsProvider sounds={sounds}>
                        <ImagesProvider images={images}>{children}</ImagesProvider>
                    </SoundsProvider>
                </ThemeProvider>
            </Provider>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return {store, ...render(ui, {wrapper: Wrapper, ...renderOptions})};
}
