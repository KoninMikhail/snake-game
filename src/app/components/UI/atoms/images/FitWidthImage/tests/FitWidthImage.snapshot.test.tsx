import { FitWidthImage } from '../FitWidthImage';
import { renderWithProviders } from '@utils/test-utils';

/**
 * Constants
 */
const IMAGE_ALT = 'app Name Here';
const IMAGE_TEST_ID = 'custom-logo';

/**
 * Tests
 */
describe('FitWidthImage', () => {
    it('equal snapshot', () => {
        const { container } = renderWithProviders(
            <FitWidthImage sourceID={'logo'} alt={IMAGE_ALT} data-testid={IMAGE_TEST_ID} />
        );

        expect(container).toMatchSnapshot();
    });
});
