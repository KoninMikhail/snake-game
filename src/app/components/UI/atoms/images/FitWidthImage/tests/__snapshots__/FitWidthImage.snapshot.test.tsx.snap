// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FitWidthImage equal snapshot 1`] = `
.c0 {
  width: 100%;
  height: auto;
}

<div>
  <img
    alt="app Name Here"
    class="c0"
    data-testid="custom-logo"
    src="PLACEHOLDER_SRC"
  />
</div>
`;
